mumd/src/audio/output.rs

```diff
@@ -1,22 +1,32 @@
 use crate::network::VoiceStreamType;
+use crate::audio::SAMPLE_RATE;
+use crate::error::{AudioError, AudioStream};
 
-use cpal::{OutputCallbackInfo, Sample};
+use log::*;
+use cpal::traits::{DeviceTrait, HostTrait, StreamTrait};
+use cpal::{SampleFormat, SampleRate, StreamConfig, OutputCallbackInfo, Sample};
 use mumble_protocol::voice::VoicePacketPayload;
 use opus::Channels;
 use std::collections::{HashMap, VecDeque};
 use std::ops::AddAssign;
 use std::sync::{Arc, Mutex};
 use tokio::sync::watch;
 
+type ClientStreamKey = Option<(VoiceStreamType, u32)>;
+type ConsumerInput<'a> = std::collections::hash_map::IterMut<'a, ClientStreamKey, VecDeque<f32>>;
+
 pub struct ClientStream {
-    buffer: VecDeque<f32>, //TODO ring buffer?
+    buffer: HashMap<ClientStreamKey, VecDeque<f32>>, //TODO ring buffer?
```

Comment:
This new ClientStream have only one opus_decoder for all users. I could not find a reason to have one each instance per client, so I just changed it.

Reply:
I haven't taken a full look at this yet but I can at least give some context here.

We had 1 decoder per user all the way from the start. I don't recall exactly why, but I _think_ it was only because we read in the [`opus::Decoder` documentation](https://docs.rs/opus/0.2.1/opus/struct.Decoder.html) that it has associated state and we didn't want state from different audio streams to mix together.

We ran into problems with this when we implemented the "Tunneling audio over TCP" part of the Mumble protocol. Sometimes, when switching to and from UDP and TCP, the audio started to sound glitched for a bit before catching up. Our conclusion (as I recall it) was that duplicate packets were being sent for a while which caused the decoder state to get confused. This is why we have 55644de. (The history is a mess. Sorry. And we should've added comments explaining this somewhere. Double sorry.)

I'll try to reproduce this with your single audio stream. If it works I think it's a better solution.